### PR TITLE
Reuse grpc connections in pachd

### LINKED ISSUE #588: pachd opens way too many gRPC connections
This happens when you write a lot of files.  The number of ports used can shoot up to 20k+.  It also causes a significant amount of memory usage.  For instance, over 5GB of memory is allocated by gRPC over the course of writing 10000+ files.

```
      flat  flat%   sum%        cum   cum%
191164050.63kB 95.74% 95.74% 191164050.63kB 95.74%  github.com/pachyderm/pachyderm/src/server/vendor/bazil.org/fuse.allocMessage
5145702.55kB  2.58% 98.32% 5179682.23kB  2.59%  github.com/pachyderm/pachyderm/src/server/vendor/google.golang.org/grpc/transport.newFramer
```

### PR DESCRIPTION
**Pull request by derekchiang**
This PR makes it so that we reuse grpc connections from API severs to internal API servers, instead of creating a new one for each request.  In my setup, under heavy load, this reduces the memory usage of pachd by about 85%.

Also fixes #588.

Here is a brief history of the evolution of our handling of grpc connections in pachd:

1. We used to reuse grpc connections similar to (but not the same as) the way it's done in this PR.  We used to only create a new connection if `conn.State()` reports that the connection is shutting down.

2. However, the `State()` API was [removed recently from grpc](https://github.com/grpc/grpc-go/pull/690).  We then resorted to creating a new connection for each request.  However, we forgot about closing the connections.

3. Later, we found out about a deadlock that happens when a pachd node leaves the cluster.  Basically, the following sequence of events could occur:

    1. A pachd node leaves the cluster
    2. In-flight requests to the node blocks while holding the `version` lock
    3. Nodes are unable to update their version because `version` lock is being held
    
    Therefore, [we introduced a fix](https://github.com/pachyderm/pachyderm/pull/464) where we cancel in-flight requests when a new version comes in.  We also realized at this point that we are not closing the connections we created, so we added `defer clientConn.Close()` everywhere.

4. Now, we found out that creating a lot of connections results in a very high memory usage.  So I took a step back and realized that there's actually no reason to create a new connection every time, even without the `State()` API.  We only needed the `State()` API because we didn't want requests to block indefinitely when the receiver (an internal API server) is down.  But now that we have a mechanism to cancel in-flight requests, it's perfectly fine to reuse connections, even if some of them may not be functioning anymore.  That is, even though in-flight requests will still block on non-functioning connections, they will now be canceled when a new version comes in.

To summarize how the handling now is different than before the removal of the `State()` API: before, we close a connection when a connection's state becomes `Shutdown`.  Now, we close all outstanding connections when the cluster version increments.

### CODE REVIEW

**Comment by jdoliner** [comment]
I'm not totally sure this works. What happens if a pachd pod restarts and comes back up with the same address? Since we never evict connection it seems like we'll be left with a defunct connection and anytime we try to use it it'll block.

I think you could test this pretty easily by creating a test similar to `TestRestartAll` which just deletes 1 of the pachd pods.

**Comment by derekchiang** [comment]
If the pod restarts and results in a membership change, then connections will be evicted.

If the pod restarts immediately without a membership changes, then grpc will simply redial.  A gRPC connection only ever shut down if either side explicitly called `Close()`.

But yeah, I can certainly add that test case.

**Comment by jdoliner** [comment]
Ahh, I didn't realize gRPC could redial. Assuming that works then this PR is lgtm. Will begin hunting for a meme.

File: src/server/pachyderm_test.go [hunk 1]
```diff
@@ -2200,3 +2241,17 @@ func restartAll(t *testing.T) {
 	}
 	waitForReadiness(t)
 }
+
+func restartOne(t *testing.T) {
+	k := getKubeClient(t)
+	podsInterface := k.Pods(api.NamespaceDefault)
+	labelSelector, err := labels.Parse("suite=pachyderm")
```
Comment (jdoliner) [inline src/server/pachyderm_test.go]: This could lead to restarting the rethink or etcd pod which I don't think tests what we want. You should do `app=pachd` instead.

**Comment by derekchiang** [comment]
@jdoliner no meme no merge

**Comment by jdoliner** [comment]
lgtm
